clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/subscriber/PubSubSubscriber.java

```diff
@@ -0,0 +1,117 @@
+import org.slf4j.LoggerFactory;
+
+import com.google.api.core.ApiService;
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+
+/**
+ * A client for receiving messages from Pub/Sub.
+ * <p>
+ * Wraps a Pub/Sub Subscriber.
+ * </p>
+ */
+public class PubSubSubscriber implements AutoCloseable {
+
+    private final Logger log = LoggerFactory.getLogger(PubSubSubscriber.class);
+    private final Subscriber subscriber;
+
+    private final Vertx vertx;
+
+    /**
+     * Creates a new instance of PubSubSubscriberClient where a Pub/Sub Subscriber is initialized. The Subscriber is
+     * based on a created subscription, which follows the format: projects/{project}/subscriptions/{subscription}
+     *
+     * @param vertx The Vert.x instance that this subscriber runs on.
+     * @param projectId The identifier of the Google Cloud Project to connect to.
+     * @param subscriptionId The name of the subscription to create the subscriber for.
+     * @param receiver The message receiver used to process the received message.
+     * @param credentialsProvider The provider for credentials to use for authenticating to the Pub/Sub service.
+     * @throws NullPointerException If any of these parameters is {@code null}.
+     */
+    public PubSubSubscriber(
+            final Vertx vertx,
+            final String projectId,
+            final String subscriptionId,
+            final MessageReceiver receiver,
+            final FixedCredentialsProvider credentialsProvider) {
+        this.vertx = Objects.requireNonNull(vertx);
+        Objects.requireNonNull(projectId);
+        Objects.requireNonNull(subscriptionId);
+        Objects.requireNonNull(receiver);
+        Objects.requireNonNull(credentialsProvider);
+
+        final ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId);
+        this.subscriber = Subscriber
+                .newBuilder(subscriptionName, receiver)
+                .setCredentialsProvider(credentialsProvider)
+                .build();
+    }
+
+    /**
+     * Subscribes messages from Pub/Sub.
+     *
+     * @return A future indicating the outcome of the operation.
+     * @throws ServerErrorException If subscribing was not successful.
+     */
+    public Future<Void> subscribe() {
+        try {
+            subscriber.addListener(
+                    new ApiService.Listener() {
+
+                        @Override
+                        public void failed(final ApiService.State from, final Throwable failure) {
+                            log.error("Error subscribing message from Pub/Sub", failure);
+                            throw new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE,
```

Comment:
The listener will have the corresponding transition method called whenever the service changes state. So this is called when the Subscriber encounters a fatal error (e.g. credentials are removed, subscription on Pub/Sub is deleted). I have changed the behavior to catch the exception & retry to subscribe every 60 seconds and also only in case this is wanted.